pkg/manager/member/pvc_resizer.go

```
@@ -80,215 +79,361 @@ type pvcResizer struct {
 	ns := tc.GetNamespace()
-	selector, err := label.New().Instance(tc.GetInstanceName()).Selector()
-	if err != nil {
-		return err
+	name := tc.GetName()
+
+	components := []v1alpha1.MemberType{}
+	if tc.Spec.PD != nil {
+		components = append(components, v1alpha1.PDMemberType)
+	}
+	if tc.Spec.TiDB != nil {
+		components = append(components, v1alpha1.TiDBMemberType)
+	}
+	if tc.Spec.TiKV != nil {
+		components = append(components, v1alpha1.TiKVMemberType)
+	}
+	if tc.Spec.TiFlash != nil {
+		components = append(components, v1alpha1.TiFlashMemberType)
+	}
+	if tc.Spec.TiCDC != nil {
+		components = append(components, v1alpha1.TiCDCMemberType)
+	}
+	if tc.Spec.Pump != nil {
+		components = append(components, v1alpha1.PumpMemberType)
 	}
 
-	// For each component, we compose a map called pvcPrefix2Quantity, with PVC name prefix for current component as keys,
-	// for example "pd-${tcName}-pd" (for tc.Spec.PD.Requests) or "pd-log-${tcName}-pd" (for tc.Spec.PD.storageVolumes elements with name "log").
-	// Reference implementation of BuildStorageVolumeAndVolumeMount().
-	// Note: for TiFlash, it is currently "data0-${tcName}-tiflash" (for tc.Spec.TiFlash.StorageClaims elements, in list definition order)
+	for _, comp := range components {
+		ctx, err := p.buildContextForTC(tc, comp)
+		if err != nil {
+			return fmt.Errorf("sync pvc of %s for tc %s/%s failed: failed to prepare: %v", comp, ns, name, err)
+		}
 
-	// patch PD PVCs
-	if tc.Spec.PD != nil {
-		pvcPrefix2Quantity := make(map[string]resource.Quantity)
-		pdMemberType := v1alpha1.PDMemberType.String()
-		if quantity, ok := tc.Spec.PD.Requests[corev1.ResourceStorage]; ok {
-			key := fmt.Sprintf("%s-%s-%s", pdMemberType, tc.Name, pdMemberType)
-			pvcPrefix2Quantity[key] = quantity
+		p.updateVolumeStatus(ctx)
+
+		err = p.resizeVolumes(ctx)
+		if err != nil {
+			return fmt.Errorf("sync pvc for tc %s/%s failed: resize volumes for %q failed: %v", ns, name, comp, err)
```

Is it better that we aggregate the errors so that failure for one component does not block the other ones?

---

`buildContextForTC` is the prerequisite so that need to return an error. Now only `resizeVolumes` maybe return error so it doesn't matter.

---

I mean that if the resize for PD fails, can we continue to resize TiKV?

---

Done